xbmc/guilib/GUIFontTTF.cpp

```diff
@@ -651,163 +1258,177 @@ std::vector<CGUIFontTTF::Glyph> CGUIFontTTF::GetHarfBuzzShapedGlyphs(const vecTe
+    hb_buffer_set_script(buffer, run.script);
+    hb_buffer_set_content_type(buffer, HB_BUFFER_CONTENT_TYPE_UNICODE);
 
-    for (unsigned int j = run.m_startOffset; j < run.m_endOffset; j++)
+    for (unsigned int j = run.startOffset; j < run.endOffset; j++)
     {
-      hb_buffer_add(run.m_buffer, static_cast<wchar_t>(0xffff & text[j]), j);
+      hb_buffer_add(buffer, text[j].letter, j);
     }
 
-    hb_buffer_set_content_type(run.m_buffer, HB_BUFFER_CONTENT_TYPE_UNICODE);
-    hb_shape(m_hbFont, run.m_buffer, nullptr, 0);
+    if (run.font)
+      hb_shape(run.font->m_hbFont, buffer, nullptr, 0);
+
     unsigned int glyphCount;
-    run.m_glyphInfos = hb_buffer_get_glyph_infos(run.m_buffer, &glyphCount);
-    run.m_glyphPositions = hb_buffer_get_glyph_positions(run.m_buffer, &glyphCount);
+    run.glyphInfos = hb_buffer_get_glyph_infos(buffer, &glyphCount);
+    run.glyphPositions = hb_buffer_get_glyph_positions(buffer, &glyphCount);
 
     for (size_t k = 0; k < glyphCount; k++)
     {
-      glyphs.emplace_back(run.m_glyphInfos[k], run.m_glyphPositions[k]);
+      glyphs.emplace_back(run.glyphInfos[k], run.glyphPositions[k], run.font);
     }
-
-    hb_buffer_destroy(run.m_buffer);
   }
+  hb_buffer_destroy(buffer);
 
   return glyphs;
 }
 
-CGUIFontTTF::Character* CGUIFontTTF::GetCharacter(character_t chr, FT_UInt glyphIndex)
+Character* CGUIFontTTF::GetCharacter(const character_t& chr,
+                                     FT_UInt glyphIndex,
+                                     CFontData* fontData)
 {
-  const wchar_t letter = static_cast<wchar_t>(chr & 0xffff);
+  typedef uint32_t char_quik_t;
+
+  const auto letter = chr.letter;
 
   // ignore linebreaks
-  if (letter == L'\r')
+  if (letter == static_cast<decltype(chr.letter)>('\r'))
     return nullptr;
 
-  const character_t style = (chr & 0x7000000) >> 24; // style = 0 - 6
+  const auto style = chr.style;
+
+  if (!fontData)
+  {
+    glyphIndex = FT_Get_Char_Index(m_fontMain.m_face, chr.letter);
+    if (glyphIndex != 0)
+      fontData = &m_fontMain;
+    else
+    {
+      const CFontTable* fontTable;
+      fontData = FindFallback(chr.letter, fontTable);
+      if (!fontData || (glyphIndex = FT_Get_Char_Index(fontData->m_face, chr.letter)) != 0)
+      {
+        CLog::Log(LOGERROR,
+                  "CGUIFontTTF::{}: Unable to find font about character {:x}, font data found: {}",
+                  __func__, static_cast<uint32_t>(chr.letter), fontData ? "yes" : "no");
+        return nullptr;
+      }
+    }
+  }
 
   if (!glyphIndex)
-    glyphIndex = FT_Get_Char_Index(m_face, letter);
+    glyphIndex = FT_Get_Char_Index(fontData->m_face, chr.letter);
 
   // quick access to the most frequently used glyphs
-  if (glyphIndex < MAX_GLYPH_IDX)
+  if (glyphIndex < MAX_GLYPH_IDX && fontData->m_mainFont)
   {
-    character_t ch = (style << 12) | glyphIndex; // 2^12 = 4096
+    char_quik_t ch = (style << 12) | glyphIndex; // 2^12 = 4096
```

Comment:
Maybe `const uint32_t` here and in the other place that is used

Reply:
Can change it, thought first to have more clean, as the "ch" not direct understandable and stores the character a bit different with this flags.

Reply:
Have changed to `const uint32_t`